server: log batch summary

RaduBerinde opened the pull request:
Replacing the now useless "node X" event with a short summary of the request
types in the batch: e.g. `1 Put`, `242 Merge`, `1 BeginTxn, 1 CPut, 1 EndTxn`.

The new method that forms this summary is generated automatically.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9464)

<!-- Reviewable:end -->

RaduBerinde committed:
server: log batch summary

Replacing the now useless "node X" event with a short summary of the request
types in the batch: e.g. `1 Put`, `242 Merge`, `1 BeginTxn, 1 CPut, 1 EndTxn`.

The new method that forms this summary is auto-generated. CreateReply is now
also generated by the same program and shares some of the counting code; it is
also more efficient: we have one `switch` where the previous implementation had
three (in the loop, in `GetInner`, and in `SetInner`).

petermattis reviewed (approved):
LGTM

bdarnell commented:
BatchRequest.String does something similar (but not quite as concisely). Do we need two ways of summarizing a batch?

Don't check in the binary file `roachpb/gen/gen`.

RaduBerinde commented:
`BatchRequest.String` shows the keys for all the requests so it can cause quite a clutter (and it stops printing after 20 requests). I figured a quick and short way to see what kind of requests are in the batch would be more helpful.

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[roachpb/gen/gen-batch-summary.go, line 16 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-KS3Ftj5gdOF_8kn14XT-r2-16:-KS5yLgEtMAspp7Tt5JG:b-lme2cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/40adda2cb88f4e029cbaf3b2ff23e2f39605a611/roachpb/gen/gen-batch-summary.go#L16)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> You can place this in the `roachpb` package and give it an `ignore` build tag that you then pass to `go run`. 
</details>
Good idea! Though I used a more specific tag - there are various files in the standard library using `+build ignore` which causes build errors.

Also renamed the files to use underscore for consistency, and moved the test to `batch_test.go`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-:-KS5xbA79V16XUPxSHH1:b-r6xxk9)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-:-KS6erStqMBjUTeE7_70:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Haven't looked at this change, but I think we should retire the existing
`BatchRequest.String()` which is just too verbose.

On Tue, Sep 20, 2016 at 10:48 AM Tamir Duberstein <notifications@github.com>
wrote:

> Review status: 0 of 5 files reviewed at latest revision, 2 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *roachpb/batch_summary.go, line 44 at r3
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-KS73F-mCi4yw9YAp_jw:-KS73F-n3Y_FKR6eBURY:b-9nfeez>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/d0c9246a6e23f9226415d62ddea6b0dc2dcdd8d7/roachpb/batch_summary.go#L44>):*
>
> // Summary prints a short summary of the requests in a batch.func (br *BatchRequest) Summary() string {
>
> this method has an awful lot of overlap with (*BatchRequest).CreateReply.
> Do you think it makes sense to try to share some of that code?
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9464>*
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9464#issuecomment-248324046>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135OXnAENzBME_YxSzHI-NCGOOeJ4jks5qr_I3gaJpZM4KA9oX>
> .
>
-- 

-- Tobias

RaduBerinde commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[roachpb/batch_summary.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-KS73F-mCi4yw9YAp_jw:-KS83ANmHwoJjDVVvjfH:b-hmn2if) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0c9246a6e23f9226415d62ddea6b0dc2dcdd8d7/roachpb/batch_summary.go#L44)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this method has an awful lot of overlap with `(*BatchRequest).CreateReply`. Do you think it makes sense to try to share some of that code?
</details>
Thanks for pointing that one out, didn't know about it. Yes, I can make that be autogenerated as well and have them share the counting code. We can also avoid the type switch in MustSetInner->SetValue by just setting the correct ResponseUnion field in each `case`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[roachpb/batch_summary.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-KS73F-mCi4yw9YAp_jw:-KS8SUa-9JPVFx3W264B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0c9246a6e23f9226415d62ddea6b0dc2dcdd8d7/roachpb/batch_summary.go#L44)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Thanks for pointing that one out, didn't know about it. Yes, I can make that be autogenerated as well and have them share the counting code. We can also avoid the type switch in MustSetInner->SetValue by just setting the correct ResponseUnion field in each `case`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> mod @tschottdorf's comment.

---

Reviewed 1 of 8 files at r1, 3 of 9 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[roachpb/gen_batch.go, line 54 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-KS8gfj302SDKbsS6Vk_:-KS8gfj302SDKbsS6Vka:b-6ckvi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bae0e86a603efa8e25bda811703da4f56551476/roachpb/gen_batch.go#L54)):*
> ```Go
> 		field := t.Field(i).Name
> 		info.name = field
> 		// The ResponseUnion fields match those in RequestUnion, with a couple of
> ```

Renaming them would be harmless, FWIW.

---

*[roachpb/gen_batch.go, line 131 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-KS8gwiIM5HfMcBodwVr:-KS8gwiIM5HfMcBodwVs:b-8a9v5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bae0e86a603efa8e25bda811703da4f56551476/roachpb/gen_batch.go#L131)):*
> ```Go
> 		"Admin":       "Adm",
> 		"Increment":   "Inc",
> 		"Conditional": "C",
> ```

CPut?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-:-KS8h3Mi1bNCvfEuL6RV:b-41vx4)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
We use `BatchRequest.String()` as part of some traces (dist_sender, senderTransport) and verbose logs (`logRaftReady`), and probably some other places I haven't identified. Are we sure we don't need the information that String outputs (which includes keys) anywhere? I think if we mess with that it should be a separate PR.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[roachpb/gen_batch.go, line 54 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-KS8gfj302SDKbsS6Vk_:-KS9A4ZQ3eqdTcc2Dnyr:blj3ih) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bae0e86a603efa8e25bda811703da4f56551476/roachpb/gen_batch.go#L54)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Renaming them would be harmless, FWIW.
</details>
One of them is not just a matter of renaming, `RequestLease *RequestLeaseResponse` is used in the response for both `RequestLease` and `TransferLease`.

---

*[roachpb/gen_batch.go, line 131 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-KS8gwiIM5HfMcBodwVr:-KS9C7MQKrRXFk2uOt-d:b-hp24pw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bae0e86a603efa8e25bda811703da4f56551476/roachpb/gen_batch.go#L131)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> CPut?
</details>
yeah it's to go from `ConditionalPut` to `CPut`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-:-KS9AIDKXwO6s9YFkMel:b-v0l8od)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I find the keys in `BatchRequest.String()` very useful. They were instrumental in figuring out #9247, for example. If we only have one, I'd rather have the one with the keys than without. I'm not sure when it's worth logging the full keys and when it's better to be concise.

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[roachpb/gen_batch.go, line 131 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464#-KS8gwiIM5HfMcBodwVr:-KS9qa-Ytzg2JkKY1a8W:b-pimmef) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bae0e86a603efa8e25bda811703da4f56551476/roachpb/gen_batch.go#L131)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> yeah it's to go from `ConditionalPut` to `CPut`.
</details>
Oh, I see what this does.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9464)*
<!-- Sent from Reviewable.io -->
